src/transport/SecureSessionMgr.cpp

```diff
@@ -111,54 +120,74 @@ CHIP_ERROR SecureSessionMgr::SendMessage(PayloadHeader & payloadHeader, NodeId p
-        VerifyOrExit(CanCastTo<uint16_t>(payloadLength), err = CHIP_ERROR_NO_MEMORY);
+
+        if (!isResend)
+        {
+            msgId = state->GetSendMessageIndex();
+
+            static_assert(std::is_same<decltype(msgBuf->TotalLength()), uint16_t>::value,
+                          "Addition to generate payloadLength might overflow");
+            payloadLength = static_cast<uint32_t>(headerSize + msgBuf->TotalLength());
+            VerifyOrExit(CanCastTo<uint16_t>(payloadLength), err = CHIP_ERROR_NO_MEMORY);
+        }
 
         packetHeader
             .SetSourceNodeId(mLocalNodeId)              //
             .SetDestinationNodeId(peerNodeId)           //
-            .SetMessageId(state->GetSendMessageIndex()) //
+            .SetMessageId(msgId)                        //
             .SetEncryptionKeyID(state->GetLocalKeyID()) //
             .SetPayloadLength(static_cast<uint16_t>(payloadLength));
         packetHeader.GetFlags().Set(Header::FlagValues::kSecure);
 
         ChipLogProgress(Inet, "Sending msg from %llu to %llu", mLocalNodeId, peerNodeId);
 
-        VerifyOrExit(msgBuf->EnsureReservedSize(headerSize), err = CHIP_ERROR_NO_MEMORY);
+        // Skip encryption process if the packet is resent by CRMP
+        if (!isResend)
+        {
+            VerifyOrExit(msgBuf->EnsureReservedSize(headerSize), err = CHIP_ERROR_NO_MEMORY);
+
+            msgBuf->SetStart(msgBuf->Start() - headerSize);
+            data     = msgBuf->Start();
+            totalLen = msgBuf->TotalLength();
 
-        msgBuf->SetStart(msgBuf->Start() - headerSize);
-        data     = msgBuf->Start();
-        totalLen = msgBuf->TotalLength();
+            err = payloadHeader.Encode(data, totalLen, &actualEncodedHeaderSize);
+            SuccessOrExit(err);
 
-        err = payloadHeader.Encode(data, totalLen, &actualEncodedHeaderSize);
-        SuccessOrExit(err);
+            err = state->GetSecureSession().Encrypt(data, totalLen, data, packetHeader, mac);
+            SuccessOrExit(err);
+            err = mac.Encode(packetHeader, &data[totalLen], kMaxTagLen, &taglen);
+            SuccessOrExit(err);
 
-        err = state->GetSecureSession().Encrypt(data, totalLen, data, packetHeader, mac);
-        SuccessOrExit(err);
+            VerifyOrExit(CanCastTo<uint16_t>(totalLen + taglen), err = CHIP_ERROR_INTERNAL);
+            msgBuf->SetDataLength(static_cast<uint16_t>(totalLen + taglen), nullptr);
 
-        err = mac.Encode(packetHeader, &data[totalLen], kMaxTagLen, &taglen);
-        SuccessOrExit(err);
+            ChipLogDetail(Inet, "Secure transport transmitting msg %u after encryption", msgId);
+        }
 
-        VerifyOrExit(CanCastTo<uint16_t>(totalLen + taglen), err = CHIP_ERROR_INTERNAL);
-        msgBuf->SetDataLength(static_cast<uint16_t>(totalLen + taglen), nullptr);
+        // Locally store the start and length;
+        p   = msgBuf->Start();
```

Comment:
I don't think this will work.  That  'p' becomes invalid when `SendMessage()` below releases the PacketBuffer.

I think you'll want to save the return value of msgBuf.Retain() and call `SendMessage()` with the return value of `Release_ForNow()`

@kpschoedel @bzbarsky-apple

Reply:
Assuming you don't need the value `p` to outlive this function somewhere, I think you want
```
err  = mTransportMgr->SendMessage(packetHeader, state->GetPeerAddress(), msgBuf.Retain());
```
Then `SendMessage` or its children will release its reference when it's done, and this function will release its reference when it returns.

Reply:
We need to rewind the start data before the encryption portion in PacketBuffer buffer for re-transmission case

"p   = msgBuf->Start();"  is equivalent to "p = msgBuf.mBuffer->Start();"

After msgBuf.Release_ForNow(), the mBuffer will be set to nullptr, but the PacketBuffer is not released here. To prove this, here is the log I captured with echo app with SetDropAckDebug(true);

CHIP:EM: ExchangeContext::HandleMessage
Echo Request from node 12344321, len=15 ... sending response.
CHIP:EM: Piggybacking Ack for MsgId:00000000 with msg
CHIP:IN: ReliableMessageManager::SendFromRetransTable: entry->msgBuf->Start()= c00d0a6e, len=15             --> first send
CHIP:IN: Sending msg from 12344322 to 12344321
CHIP:IN: Secure transport transmitting msg 0 after encryption
CHIP:IN: SecureSessionMgr::SendMessage: before msgBuf.Release_ForNow(): p= c00d0a64, len=41                 --> backup buffer start data
CHIP:IN: SecureSessionMgr::SendMessage: after msgBuf.Release_ForNow(): p= c00d0a64, len=41                  --> reset buffer start data
CHIP:EM: entry is updated to payloadLen = 25. MsgId = 00000000
CHIP:EM: ExchangeContext::Close
CHIP:EM: ExchangeContext::DoClose

CHIP:IN: ReliableMessageManager::ResendFromRetransTable: entry->msgBuf->Start()= c00d0a64, len=41            --> first CRMP retry, the buffer start date is valid.
CHIP:IN: Sending msg from 12344322 to 12344321
CHIP:IN: SecureSessionMgr::SendMessage: before msgBuf.Release_ForNow(): p= c00d0a64, len=41
CHIP:IN: SecureSessionMgr::SendMessage: after msgBuf.Release_ForNow(): p= c00d0a64, len=41
CHIP:EM: entry is updated to payloadLen = 25. MsgId = 00000000

CHIP:EM: Retransmit MsgId:00000000 Send Cnt 2
CHIP:IN: ReliableMessageManager::ResendFromRetransTable: entry->msgBuf->Start()= c00d0a64, len=41            --> second CRMP retry, the buffer start date is valid.
CHIP:IN: Sending msg from 12344322 to 12344321
CHIP:IN: SecureSessionMgr::SendMessage: before msgBuf.Release_ForNow(): p= c00d0a64, len=41
CHIP:IN: SecureSessionMgr::SendMessage: after msgBuf.Release_ForNow(): p= c00d0a64, len=41
CHIP:EM: entry is updated to payloadLen = 25. MsgId = 00000000

CHIP:EM: Retransmit MsgId:00000000 Send Cnt 3
CHIP:EM: Failed to Send CHIP MsgId:00000000 sendCount: 3 max retries: 3
CHIP:EM: ExchangeContext::Free
CHIP:EM: ExchangeContext::DoClose

Reply:
> Assuming you don't need the value `p` to outlive this function somewhere, I think you want
> 
> ```
> err  = mTransportMgr->SendMessage(packetHeader, state->GetPeerAddress(), msgBuf.Retain());
> ```
> 
> Then `SendMessage` or its children will release its reference when it's done, and this function will release its reference when it returns.

The buffer need to be retained in the messaging layer instead of transport layer. Since transport layer has not knowledge if the buffer is sent only for one time or need to be retained for transmitting later.  

In our case, if the message is sent from transmission table, we need to retain it.  Otherwise, it will be released by transport layer after sending.

CHIP_ERROR ReliableMessageManager::SendFromRetransTable(RetransTableEntry * entry, bool isResend)
{
 .......
       msgBuf.Adopt(entry->msgBuf);

        // Retain entry->msgBuf in mRetransTable for following retransmition
        msgBuf->AddRef();
.......
}

In summary, no logic change in transport layer, it only needs to skip the encryption process if the message is resent from upper layer since the payload has already been encrypted.

Reply:
the facility you need to use is to increment the refcount on the underlying pbuf.  the code you have here does not accomplish that.

Reply:
`msgBuf.Retain()` provides a new PacketBufferHandle with shared ownership of the buffer. `SendMessage()` or its children will drop that ownership when they no longer need the buffer, and likewise this can code drop its ownership (held by `msgBuf`) when it no longer needs it.

`AddRef()` will soon be private.

Reply:
I guess the issue that the messaging layer doesn't know the right boundaries to set, because those will change in this function?  But the messaging layer is what knows we're retaining?

I think we have a few options here:

1.  We could have this function take an optional argument for a place to return the start and length of the buffer as it was after we finished setting it up.  The callers who care about it would pass a non-null pointer there.
2. We could have this function take an optional `PacketBufferHandle*` pointing to a handle that is going to retain the buffer after this function call, and make its calls on that.  This seems a bit error-prone if caller passes a pointer to the thing it's moving out of....
3.  The caller could pass in a boolean indicating "I am retaining this buffer" and then we can just set state on the buffer or something.  Also seems error-prone.
4.  The caller could pass in a non-null pointer to a PacketBufferHandle to indicate that: we should (a) Retain() the buffer into that object and (b) preserve its start/length.  I sort of like this approach: it centralizes the "we are retaining the buffer" and "we are retaining its state" logic in a single place.  It even works correctly if the caller uses a pointer to the same handle that it `std::move`s into our "this is the data to sent" argument.

Open to other suggestions too.  Whatever we decide on, the API documentation needs to document it clearly.

Reply:
Changed to increment the refcount on System::PacketBuffer directly
entry->msgBuf->AddRef();

We store the raw System::PacketBuffer in RetransTable, and  use PacketBufferHandle as the wrapper to sent this packet to Transport layer, even the ownership of PacketBufferHandle is transferred, but the raw System::PacketBuffer, should not be freed at the lower layer if we increment the reference counter here.

The retained System::PacketBuffer will be recycled when it is cleared from RetransTable within void ReliableMessageManager::ClearRetransmitTable(RetransTableEntry & rEntry).
System::PacketBuffer::Free(rEntry.msgBuf);;

Reply:
> Changed to increment the refcount on System::PacketBuffer directly

Why?  Again, `AddRef` is slated to become private soon, so it's best to not add new calls to it.

> We store the raw System::PacketBuffer in RetransTable

Why?  Why does the table not store `PacketBufferHandle`?  It should.

Reply:
What we need to achieve is to retain the  System::PacketBuffer  after it is added to RetransTable for the following re-transmit, RetransTable will hold this buffer until it is acked or expire the maximum re-transmit. 

If we hold  PacketBufferHandle in Retransmit, the ownership of this packet will be transferred to Transport layer after each send, how to handle the following re-transmit case? Are we going to make AddRef private for PacketBufferHandle or System::PacketBuffer? I think only to PacketBufferHandle, right?

Reply:
The ownership of PacketBufferHandle will be transferred to Transport layer after each send, right?

    PacketBufferHandle(PacketBufferHandle && aOther)
    {
        mBuffer        = aOther.mBuffer;
        aOther.mBuffer = nullptr;
    }

We just want to retain the System::PacketBuffer in RetransTable until it is acked or expired. Do we plan to make AddRef private for PacketBufferHandle only or also to System::PacketBuffer?

Reply:
> The ownership of PacketBufferHandle will be transferred to Transport layer after each send, right?

Depends on how you send it.  E.g.:
```
Send(myTable[i].myHandle.Retain());
```
would keep holding on to the thing in the table and give the transport layer its own reference to manage.

> Do we plan to make AddRef private for PacketBufferHandle only or also to System::PacketBuffer?

The latter.

Reply:
Yes, PacketBufferHandle can be co-owned by multiple owners with function Retain(), updated to use PacketBufferHandle to store message buffer in re-transmit table

Reply:
Just to be clear, a `PacketBufferHandle` always has a unique owner, with ownership transfer possible via `std::move`.  But multiple handles can be pointing to the same `PacketBuffer` and all keeping it alive, and `Retain` creates a new such handle pointing to the same `PacketBuffer` as the handle `Retain` was called on.